### PR TITLE
build a `skills/requests` api & create a rds service to get user details

### PR DESCRIPTION
**Pull request by yesyash**
## Issue Ticket Number
- 137

### Description
- create a `RdsService` to get user data from the rds backend.
- create a `v1/skills/requests` API to get all pending skills requests on the platform.
    - this API is only accessible to the superusers.
    - the API structure can be seen in the sample response below.
- add `RDS_BACKEND_BASE_URL` key to env & application properties.

### Development Tested?
- [x] Yes
- [ ] No

## Screenshots
image

#### verification of two requests being made to the rds backend as we only have two users' data required in the user's array.
image

### CODE REVIEW

File: skill-tree/src/main/java/com/RDS/skilltree/services/SkillServiceImplementation.java [hunk 1]
```diff
@@ -30,6 +37,71 @@ public List<SkillViewModel> getAll() {
                 .collect(Collectors.toList());
     }
 
+    @Override
+    public SkillRequestsDto getAllRequests() {
```
Comment (iamitprakash) [inline skill-tree/src/main/java/com/RDS/skilltree/services/SkillServiceImplementation.java:41]: 1. Move the logic for fetching user details outside the map and forEach loop.
2. Create a set to store user IDs for whom details have already been fetched.
3. Before fetching details using rdsService.getUserDetails, check if the user ID is already present in the set.
4. If the ID is not present in the set, fetch the details, add the ID to the set and store the details in the userDetails map.

File: skill-tree/src/main/java/com/RDS/skilltree/services/external/RdsServiceImplementation.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+@Service
+@RequiredArgsConstructor
+public class RdsServiceImplementation implements RdsService {
+    private final RestTemplate restTemplate;
+
+    @Value("${rds.backendBaseUrl}")
+    private String rdsBackendBaseUrl;
+
+    @Override
+    public RdsUserViewModel getUserDetails(String id) {
```
Comment (iamitprakash) [inline skill-tree/src/main/java/com/RDS/skilltree/services/external/RdsServiceImplementation.java:18]: can we wrap this into a try-catch block to handle exceptions and return appropriate error responses or log the issue for debugging.

File: skill-tree/src/main/java/com/RDS/skilltree/viewmodels/RdsUserViewModel.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    @Setter
+    public static class User {
+        private String id;
+        private boolean incomplete_user_details;
+        private String discord_joined_at;
+        private String discord_id;
+        private Roles roles;
+        private String linkedin_id;
+        private Picture picture;
+        private int yoe;
```
Comment (iamitprakash) [inline skill-tree/src/main/java/com/RDS/skilltree/viewmodels/RdsUserViewModel.java:23]: what about yoe= 2.4

File: skill-tree/src/main/java/com/RDS/skilltree/services/external/RdsServiceImplementation.java (hunk 2 above)
Comment (iamitprakash) [inline skill-tree/src/main/java/com/RDS/skilltree/services/external/RdsServiceImplementation.java:18]: Here url is being constructed by string concatenation. This approach can be vulnerable to security issues like https://en.wikipedia.org/wiki/SQL_injection if id is not properly sanitized.
Consider using a parameterized query with the restTemplate.

File: skill-tree/src/main/java/com/RDS/skilltree/viewmodels/SkillRequestViewModel.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+@NoArgsConstructor
+public class SkillRequestViewModel {
+    private Integer skillId;
+    private String skillName;
+    private String endorseId;
+    private List<MinimalEndorsementViewModel> endorsements;
+
+    public static SkillRequestViewModel toViewModel(
+            UserSkillsModel userSkillsModel, List<Endorsement> endorsements) {
+        SkillRequestViewModel skillRequestViewModel = new SkillRequestViewModel();
```
Comment (iamitprakash) [inline skill-tree/src/main/java/com/RDS/skilltree/viewmodels/SkillRequestViewModel.java:22]: here a new SkillRequestViewModel object is created and then its fields are populated. You can potentially simplify this by using a constructor or setter methods in SkillRequestViewModel that takes the required parameters. *Redundant object creation*

can we use something like below 
public static SkillRequestViewModel toViewModel(UserSkillsModel userSkillsModel, List<Endorsement> endorsements) {
  Skill skill = userSkillsModel.getSkill();
  UserModel user = userSkillsModel.getUser();
  
  return new SkillRequestViewModel(
      skill.getId(),
      skill.getName(),
      user.getId(),
      endorsements.stream()
          .map(MinimalEndorsementViewModel::toViewModel)
          .collect(Collectors.toList()));
}

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: skill-tree/src/main/java/com/RDS/skilltree/viewmodels/RdsUserViewModel.java (hunk 3 above)
Comment (yesyash) [reply skill-tree/src/main/java/com/RDS/skilltree/viewmodels/RdsUserViewModel.java:23]: - changed this to `float`

File: skill-tree/src/main/java/com/RDS/skilltree/services/external/RdsServiceImplementation.java (hunk 2 above)
Comment (yesyash) [reply skill-tree/src/main/java/com/RDS/skilltree/services/external/RdsServiceImplementation.java:22]: - added a try-catch block

File: skill-tree/src/main/java/com/RDS/skilltree/viewmodels/SkillRequestViewModel.java (hunk 4 above)
Comment (yesyash) [reply skill-tree/src/main/java/com/RDS/skilltree/viewmodels/SkillRequestViewModel.java:22]: - created a constructor and directly returned a the object after creating it like you suggested:
```java
return new SkillRequestViewModel(
skill.getId(),
skill.getName(),
user.getId(),
endorsements.stream()
.map(MinimalEndorsementViewModel::toViewModel)
.collect(Collectors.toList()));
}
```

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**